kubeadm: fix offline and air-gapped support

neolit123 opened the pull request:
**What this PR does / why we need it**:

1.

Change the error output of getAllDefaultRoutes() so that it includes
information on which files were probed for the IP routing tables
even if such files are obvious.

Introduce a new error type which can be used to figure out of this
error is exactly of the "no routes" type.

2.

If netutil.ChooseBindAddress() fails looking up IP route tables
it will fail with an error in which case the kubeadm config
code will hard stop.

This scenario is possible if the Linux user intentionally disables
the WiFi from the distribution settings. In such a case the distro
could empty files such files as /proc/net/route and ChooseBindAddress()
will return an error.

For improved offline support, don't error on such scenarios but instead
show a warning. This is done by using the NoRoutesError type.
Also default the address to 0.0.0.0.

While doing that, prevent some commands like `init`, `join` and also
phases like `controlplane` and `certs` from using such an invalid
address.

3.

If there is no internet, label versions fail and this breaks
air-gapped setups unless the users pass an explicit version.

To work around that:
- Remain using 'release/stable-x.xx' as the default version.
- On timeout or any error different from status 404 return error
- On status 404 fallback to using the version of the client via
kubeadmVersion()

Add unit tests for kubeadmVersion().

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
refs kubernetes/kubeadm#1041

**Special notes for your reviewer**:
1st and second commits fix offline support.
3rd commit fixes air-gabbed support (as discussed in the linked issue)

the api-machinery change is only fmt.Errorf() related.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kubeadm: fix air-gapped support and also allow some kubeadm commands to work without an available networking interface
```

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews 
/cc @kubernetes/sig-api-machinery-pr-reviews 
/assign @kad
/assign @xiangpengzhao 
/area UX
/area kubeadm
/kind bug

neolit123 committed:
pkg/util/net: use a more descriptive error in getAllDefaultRoutes()

Change the error output of getAllDefaultRoutes() so that it includes
information on which files were probed for the IP routing tables
even if such files are obvious.

Introduce a new error type which can be used to figure out of this
error is exactly of the "no routes" type.

neolit123 committed:
kubeadm: fix the air-gapped and offline support issues

1) Do not fail in case a bind address cannot be obtained

If netutil.ChooseBindAddress() fails looking up IP route tables
it will fail with an error in which case the kubeadm config
code will hard stop.

This scenario is possible if the Linux user intentionally disables
the WiFi from the distribution settings. In such a case the distro
could empty files such files as /proc/net/route and ChooseBindAddress()
will return an error.

For improved offline support, don't error on such scenarios but instead
show a warning. This is done by using the NoRoutesError type.
Also default the address to 0.0.0.0.

While doing that, prevent some commands like `init`, `join` and also
phases like `controlplane` and `certs` from using such an invalid
address.

Add unit tests for the new function for address verification.

2) Fallback to local client version

If there is no internet, label versions fail and this breaks
air-gapped setups unless the users pass an explicit version.

To work around that:
- Remain using 'release/stable-x.xx' as the default version.
- On timeout or any error different from status 404 return error
- On status 404 fallback to using the version of the client via
kubeadmVersion()

Add unit tests for kubeadmVersion().

Co-authored-by: Alexander Kanevskiy <alexander.kanevskiy@intel.com>

neolit123 committed:
kubeadm: update auto-generated BUILD files

chuckha commented on cmd/kubeadm/app/util/config/masterconfig.go:
```diff
@@ -56,9 +56,12 @@ func SetInitDynamicDefaults(cfg *kubeadmapi.InitConfiguration) error {
 	ip, err := netutil.ChooseBindAddress(addressIP)
 	if err != nil {
-		return err
+		defaultBind := "127.0.0.1"
```
The only problem I see with this is that ChooseBindAddress returns more than one error. We should really be returning a specific error type and only defaulting on the error type we expect otherwise we might ignore an actual error.

neolit123 replied:
thank makes sense.
but given it returns a string should we just string compare for "no default routes", like the unit tests are doing?

chuckha replied:
I don't like comparing strings for errors, I'd prefer type checking as described https://blog.golang.org/error-handling-and-go

neolit123 replied:
ok, i will add the error type check.

neolit123 commented:
update:
- include a custom error type so that we can verify if the error type is exactly about "no routes"

kad commented on cmd/kubeadm/app/util/config/masterconfig.go:
```diff
@@ -56,9 +56,17 @@ func SetInitDynamicDefaults(cfg *kubeadmapi.InitConfiguration) error {
+		case *netutil.NoRoutesError:
+			defaultBind := "127.0.0.1"
+			glog.Warningf("could not obtain advertise address for the API Server: %v; using: %s", err, defaultBind)
+			cfg.API.AdvertiseAddress = defaultBind
```
this will lead to the error. net space inside container with apiserver will be not the same as host, thus if we can't get bind ip, init must fail. What we can do, is to try to keep AdvertiseAddress empty, and in places where it is important, do the checks for it to have some normal values.

neolit123 replied:
@kad i was hesitant about adding 127.0.0.1.
i will give `""` a try.

kad replied:
but please have a check that it is valid in the places where it is really used. e.g. with https://golang.org/pkg/net/#IP.IsGlobalUnicast

neolit123 commented:
@kad 
updated, but i went for "0.0.0.0" for a couple of reasons:
- if we pass `""` this will fail this validation:
```
advertiseAddress: Invalid value: "": ip address is not valid
```
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/validation/validation.go#L303
and none of the commands that we want to work offline would work

- the api server uses "0.0.0.0" is the default address for the network interface as per:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L380-L382
- but even if we pass "0.0.0.0" in the config we don't need it to be a valid IP for the api server as the api server is not going to be used by the offline commands, at all. i guess this makes "127.0.0.1" also valid.

xiangpengzhao commented:
The solution SGTM :+1:

kad commented:
@neolit123  Here is example of what I've meant with checking of validity of undetected AdvertiseAddress:

```
kad:/work/go/src/k8s.io/kubernetes # _output/bin/kubeadm alpha phase certs apiserver
W0815 17:24:01.245638   28981 masterconfig.go:62] could not obtain advertise address for the API Server: no default routes found in "/proc/net/route" or "/proc/net/ipv6_route"; using: 0.0.0.0
[certificates] Generated apiserver certificate and key.
[certificates] apiserver serving cert is signed for DNS names [kad kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local] and IPs [10.96.0.1 0.0.0.0]
kad:/work/go/src/k8s.io/kubernetes # openssl x509 -in /etc/kubernetes/pki/apiserver.crt -text | grep DNS
                DNS:kad, DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, IP Address:10.96.0.1, IP Address:0.0.0.0
kad:/work/go/src/k8s.io/kubernetes # 
```
That's an error situation. If AdvertiseAddress is not detected nor specified manually in config, code that tries to use it (certificate creation, init) should fail.

neolit123 commented:
@kad

i will make the warning to be more clear:
`WARNING: could not obtain advertise address...`
to make it more clear?

> That's an error situation. If AdvertiseAddress is not detected nor specified manually in config, code that tries to use it (certificate creation, init) should fail.

i will add the check to the cert phase.

but it becomes a question of what command should fail, because things like `kubeadm config print-default` should arguably not fail even if using `0.0.0.0`, which is not exactly a valid address and can render a cluster unusable of someone uses the config as is.

kad commented:
In print defaults I think this field should be set to empty if it is 0, to force user to set proper value there.
As for which commands should fail: in my opinion, anything that does real actions and can render cluster to unusable state should fail if this IP is incorrect. This means certs and manifests generations as minimum.

dixudx commented on cmd/kubeadm/app/util/config/masterconfig.go:
```diff
@@ -56,9 +56,17 @@ func SetInitDynamicDefaults(cfg *kubeadmapi.InitConfiguration) error {
 	ip, err := netutil.ChooseBindAddress(addressIP)
 	if err != nil {
-		return err
+		switch err.(type) {
```
Shall we make it to a private function, since it is used in both masterconfig and nodeconfig.

fedebongio commented:
/assign @cheftako

neolit123 commented:
@kad updated and also included a patch for the air-gapped fix we talked about (on timeout, fallback to local client version).

neolit123 commented:
/test pull-kubernetes-integration

neolit123 commented:
/test pull-kubernetes-integration

neolit123 commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-100-performance

sttts commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:
```diff
@@ -347,7 +356,9 @@ func getAllDefaultRoutes() ([]Route, error) {
 	routes = append(routes, v6Routes...)
 	if len(routes) == 0 {
-		return nil, fmt.Errorf("No default routes.")
+		return nil, &NoRoutesError{
```
usually errors are not returned as pointers.

sttts commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:
```diff
@@ -53,6 +53,15 @@ type RouteFile struct {
 }
 
+// NoRoutesError can be returned by ChooseBindAddress() in case of no routes
+type NoRoutesError struct {
```
Would prefer to have a private type and add a `IsRoutesError` func instead.

kad commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -24,17 +24,20 @@ import (
 
 const (
-	getReleaseVersionTimeout = time.Duration(10 * time.Second)
+	getReleaseVersionTimeout = time.Duration(4 * time.Second)
```
10 seconds are actually good enough for cases with broken proxy+firewalls, which are seen as "connection timeout".
4 seconds can be too low in case of slow hardware (raspberry) + remote location (Australia, China).
In my test environment, even on good HW and quite big channels, from China it takes about 3 seconds for `curl` to get first stage HTTP reply. It is needed to be kept in mind, that `dl.k8s.io` first return redirect (HTTP 302), and then there is second "behind the scenes" http call to `https://storage.googleapis.com/kubernetes-release/....` to get actual content of label file.

neolit123 replied:
4 might be too low, indeed.
10 seems fine but the problem is that air-gapped users would have to wait "a bit" on **each** kubeadm command.
i will change it back.

kad replied:
This timeout only will be hit in the scenarios where air gaped environment has firewall that silently drop packets. In that case, yes, it will be 10 seconds waiting period. In the scenarios where firewall is actively denying connections or there is no route from the environment to the internet, after receiving ICMP "unreachable" packets, http client will immediately produce error.

neolit123 replied:
makes sense. will change it back to 10 after we clear the discussion.

i was also browsing some bash scripts today (forgot where in k8s) and i saw some code that does a `curl` on the redirected address like:
`https://storage.googleapis.com/kubernetes-release/release/latest-1.11.txt`
instead of:
`https://dl.k8s.io/release/latest-1.11.txt`

do you think we can do that to help slow connections?

kad commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -72,11 +76,19 @@ func KubernetesReleaseVersion(version string) (string, error) {
 		if err != nil {
-			return "", err
+			// Handle air-gapped environments by falling back to the client version.
+			glog.Infof("could not fetch a Kubernetes version from the internet: %v", err)
```
That fallback is not good. We should go to this fallback only in cases where we have technical issues to reach download server (timeouts, no-route-to-host, DNS failure, etc). If it was able to connect to the server, but requested label is not found (HTTP error 404), then this error about incorrect label should be propagated up.

neolit123 replied:
i agree that 404 should be handled here.

one problem, in an air-gapped scenario passing "foo" would fail fetching the version from the server with an error different from 404, this would then fallback to the client version but we don't validate a client version locally in any way and "foo" would end up being a valid version and this is bogus.

thus i think the local version should be handled with this regexp when the fallback to the client version happens:
`^(latest|stable)(-([0-9]+)(\.([0-9]+))?)?$`

from the previous comment about the regexp:
> We don't want to modify that piece of code every time when somebody creates new label on release server.

i don't see this being a big problem. if the servers change or labels change we would have to make changes regardless.

kad replied:
Wait. There is actually another issue here in your patch :)
So, for cases with fallback (result of `kubeadmVersion()` function from your code), it should be assigned to the `body` variable and let it re-validate below via recursive execution of this function. It shouldn't return immediately.

Regarding airgap issues: with 404 it is clear answer, return error. In all other scenarios (timeout, HTTP error code 5xx, etc) - fallback. We just need to add test cases for those HTTP replies in the network test piece for this code:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/version_test.go#L115

kad commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -153,3 +166,26 @@ func fetchFromURL(url string, timeout time.Duration) (string, error) {
+
+// kubeadmVersion returns the version of the client without metadata.
+func kubeadmVersion() (string, error) {
+	kubeadmVersionInfo := version.Get()
```
Result of `version.Get()` should be argument to that function. This way we can have proper tests for `kubeadmVersion()` function, which is missing now.

kad commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -24,17 +24,20 @@ import (
 	kubeReleaseBucketURL  = "https://dl.k8s.io"
 	kubeReleaseRegex      = regexp.MustCompile(`^v?(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)([-0-9a-zA-Z_\.+]*)?$`)
-	kubeReleaseLabelRegex = regexp.MustCompile(`^[[:lower:]]+(-[-\w_\.]+)?$`)
+	kubeReleaseLabelRegex = regexp.MustCompile(`^(latest|stable)(-([0-9]+)(\.([0-9]+))?)?$`)
```
I don't see need to change that regexp. Especially to put it so strict with only `latest` or `stable`. We don't want to modify that piece of code every time when somebody creates new label on release server.

neolit123 replied:
so there is a problem here with that and the fallback.

this integration test fails without the regexp local validation change:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/test/cmd/init_test.go#L92-L94

the reason is that when `foo` is passed the version is checked from the internet and when it fails (as there is no `foo` label) this falls back to the local client version and it succeeds. the 404 handling proposal should fix that.

kad replied:
I think it will be better to fix the test case linked above. This particular test case will be failing on `foobar` value based on absence on the internet or in case of network problems. So, it will be a lot better to have validation if it passes something that is clearly as incorrect semver (e.g. `1.2`) rather than something that looks like label and potentially trigger network operations during the test execution.
 
For validation of those regexps and network functions, we have separate test cases:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/version_test.go#L100
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/version_test.go#L82

So, I'd prefer to change value in https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/test/cmd/init_test.go#L92-L94 than modifying logic and regexps in `cmd/kubeadm/app/util/version.go`

kad reviewed (changes requested):
Missing things:
- too small timeout
- tests for new functions
- API server bind IP address check for IPv4 and IPv6 using standard library function
- potentially, default bind address should be set to net.IPv6zero in case cluster is going to be in IPv6 mode.

neolit123 commented on cmd/kubeadm/app/util/config/common.go:
```diff
@@ -142,3 +144,30 @@ func LowercaseSANs(sans []string) {
+// VerifyAPIServerBindAddress can be used to verify if a bind address for the API Server is 0.0.0.0,
+// in which case this address is not valid and should not be used.
+func VerifyAPIServerBindAddress(address string) error {
+	if address == constants.DefaultAPIServerBindAddress {
```
without my changes there are also scenarios where kubeadm also doesn't handle unicast addresses.

it's only handled if `ChooseBindAddress()` receives `nil` and that happens in `nodeconfiguration.go`
`https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go`
and then if this scenario is reached:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L284

but adding such a check here won't be much of a problem.

kad replied:
Yes, my point is that if we are touching this piece of code inside kubeadm, let's make sure that all the input (defaults or something that comes from user config) is validated properly. netutil does it job for choosing addresses on interfaces, but for us it is good for validate alternative input possibilities.

neolit123 replied:
ok :+1:

neolit123 commented:
thanks for the reviews, added a few comments that need minor discussion.

kad commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -138,6 +150,7 @@ func splitVersion(version string) (string, string, error) {
 
 // Internal helper: return content of URL
 func fetchFromURL(url string, timeout time.Duration) (string, error) {
+	glog.V(1).Infof("fetching Kubernetes version from URL: %s", url)
```
Do we really want to have that with V(1)? maybe with bigger verbosity level ?

neolit123 replied:
does 2 sound good?

neolit123 commented:
@kad thanks for the comments again. since the review is getting quite messy i'm just going to update based on the feedback we have already. :+1: 

soon™

kad commented:
@neolit123 let me help you a bit, #67700 for fixing test case :) I can later today refactor version fallbacks for network related issues.

neolit123 commented:
@kad
> I can later today refactor version fallbacks for network related issues.

let me have a poke at that instead because separate PRs tend to block. :\

neolit123 commented:
update:
- PR rebased against the latest kubeadm config changes
- addressed feedback by @sttts for the apimachinery change
- addressed feedback by @kad for the kubeadm change

@timothysc PTAL too

neolit123 commented:
@kad i've added more tests. :+1:

kad commented:
@neolit123 two more things:
1. Important: in cases like `v1.11.3-beta.....`, it is better to fallback to stable released version: `v1.11.2`
2. If we were able able to connect to download server, but for some reason it returned non-OK state (404 or something else), then it is not an air-gapped environment, so let's raise that error.

Both changes are in PR to your branch:  neolit123/kubernetes#1

neolit123 commented:
@kad
thanks, merged the changes.

kad commented:
@neolit123 Oops. looks like typo in one of my patches :( can you fix it within your branch ? 
I think you can also do rebase (to get rid of merge commit).

neolit123 commented:
> Oops. looks like typo in one of my patches :( can you fix it within your branch ?
I think you can also do rebase (to get rid of merge commit).

@kad 
ok i will, please have a look at the two comments above `v1.9.11-beta.0`- > `v1.9.10`

kad commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -290,3 +290,93 @@ func TestNormalizedBuildVersionVersion(t *testing.T) {
+		{
+			name:   "valid patch version with label",
+			input:  "v1.9.11-beta.0",
+			output: "v1.9.10",
```
Rationale: our default is `stable-1.x`, so as user I would expect that I'll get some stable version from the same branch as kubeadm, not random build. This will be consistent with default, and for kubeadm builds like `1.x.0*`, where the best `stable` are latest betas or latest release candidates, until it released. This way allows user by default try to install latest stable released version that is not newer than kubeadm.

neolit123 replied:
ok, we can leave it the way it is. hopefully we don't confuse the users.

kad commented:
/lgtm

neolit123 commented:
rebased after conflicts with master.

timothysc commented on cmd/kubeadm/app/util/config/common.go:
```diff
@@ -144,3 +146,32 @@ func LowercaseSANs(sans []string) {
+		}
+		return nil, err
+	}
+	*valueToUpdate = ip.String()
```
The semantics are weird of this signature, you already return the ip where any consumer can call ip.String()

timothysc replied:
I'd either call that externally or return the string.

neolit123 replied:
i will update this.

timothysc commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -138,18 +153,54 @@ func splitVersion(version string) (string, string, error) {
+	// including the offset e.g. `alpha.0.206`. This is done to comply with GCR image tags.
+	pre := v.PreRelease()
+	patch := v.Patch()
+	if len(pre) > 0 {
```
@sttts don't we have utils for this stuff somewhere?

timothysc commented on cmd/kubeadm/app/util/config/common.go:
```diff
@@ -144,3 +146,32 @@ func LowercaseSANs(sans []string) {
+
+// VerifyAPIServerBindAddress can be used to verify if a bind address for the API Server is 0.0.0.0,
+// in which case this address is not valid and should not be used.
+func VerifyAPIServerBindAddress(address string) error {
```
Are there negative tests for this?

neolit123 replied:
tests are missing. will add them.

timothysc commented:
@neolit123 - Could you squash some of the commits?

neolit123 commented:
@timothysc 
> Could you squash some of the commits?

squashed two - kubeadm related.
left the auto-generated commit separate and the one for api-machinery separate too.

p.s. i do not agree with the kubernetes commit squash policies. ;)

timothysc reviewed (approved):
/approve 

LGTM , but I'd like @kad todo final lgtm.

kad commented:
/lgtm

neolit123 commented:
looks like a verify timeout.

/test pull-kubernetes-verify
/test pull-kubernetes-e2e-gce

timothysc commented:
@kubernetes/sig-api-machinery-pr-reviews - this requires your approval.

dixudx reviewed (approved):
/lgtm

Thanks @neolit123

neolit123 commented:
@sttts could you please take a look at the api machinery change for approval.
thank you.

sttts commented:
/retest

sttts commented:
/approve
